### PR TITLE
Convolve based on `_fir_filter`

### PR DESCRIPTION
**Pull request by mgrub**
Provide a convolution method with full covariance propagation from inputs to output. Based on the `PyDynamic.uncertainty.propagate_filter._fir_filter`-method. Test compare the result to the equivalent functions from numpy/scipy. Also, a rough Monte-Carlo-simulation is performed, to check plausibility of the calculated covariance matrix.

### CODE REVIEW

**Comment by mgrub** [comment]
Consider to include further specializations of `mode=="same"`, as used in [`scipy.ndimage.convolve1d`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.convolve1d.html). These special modes are reflect, constant, nearest, mirror, wrap. 

Note: `numpy.convolve(..., mode="same")` is equivalent to `scipy.ndimage.convolve1d(..., mode="constant", cval=0.0)`
(see: [numpy.convolve](https://numpy.org/doc/stable/reference/generated))

File: test/test_propagate_convolution.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
+"""Test PyDynamic.uncertainty.propagate_convolve"""
```
Comment (BjoernLudwigPTB) [inline test/test_propagate_convolution.py:1]: These are awesomely careful chosen test cases. I would have loved to see hypothesis being facilitated to introduce the randomness, but these hand-crafted test cases are really covering all execution branches and you know best, if these touch corner cases as well. Things like the `random_array()` function are built into Hypothesis and come along with quite some intelligence for choosing the actual numerical values to discover erroneous corner cases.

File: test/test_propagate_convolution.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    numpy_modes = ["full", "valid", "same"]
+
+    if kind == "all":
+        return numpy_modes + scipy_modes
+    elif kind == "scipy":
+        return scipy_modes
+    elif kind == "numpy":
+        return numpy_modes
+    else:
+        return []
```
Comment (BjoernLudwigPTB) [inline test/test_propagate_convolution.py:65]: I feel, we should work with sets here, which is slightly more efficient with those `in` operations:
```suggestion
    scipy_modes = {"nearest", "reflect", "mirror"}
    numpy_modes = {"full", "valid", "same"}

    if kind == "all":
        return numpy_modes.union(scipy_modes)
    elif kind == "scipy":
        return scipy_modes
    elif kind == "numpy":
        return numpy_modes
    else:
        return set()
```

File: test/test_propagate_convolution.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+    # calculate the convolution of x1 and x2
+    y, Uy = convolve_unc(*input_1, *input_2, mode)
+
+    if mode in numpy_modes:
+        y_ref = np.convolve(input_1[0], input_2[0], mode=mode)
+    elif mode in scipy_modes:
+        y_ref = sn.convolve1d(input_1[0], input_2[0], mode=mode)
+    else:
+        raise ValueError(f"Unsupported Mode <{mode}>")
```
Comment (BjoernLudwigPTB) [inline test/test_propagate_convolution.py:84]: I somehow do not like the idea of assuming the test case is wrong and throwing an exception in a test case itself. One basic assumption is the test must be correct. This else-branch relies on quite some assumptions, mainly we know that any valid mode needs to be part of either what we expect to get from `valid_modes("scipy")` and `valid_modes("numpy")`. I would rather strip this logic from the test. Anyway this is what probably happens when calling Numpy or Scipy with a wrong mode, isn't it?

File: test/test_propagate_convolution.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    n_runs = 20000
+    XX1 = np.random.multivariate_normal(*input_1, size=n_runs)
+    XX2 = np.random.multivariate_normal(*input_2, size=n_runs)
+    for x1, x2 in zip(XX1, XX2):
+        if mode in numpy_modes:
+            conv = np.convolve(x1, x2, mode=mode)
+        elif mode in scipy_modes:
+            conv = sn.convolve1d(x1, x2, mode=mode)
+        else:
+            raise ValueError(f"Unsupported Mode <{mode}>")
```
Comment (BjoernLudwigPTB) [inline test/test_propagate_convolution.py:114]: same as above

**Review by BjoernLudwigPTB — COMMENTED** [review]
This is some great work and very carefully testing included as well. I added another test case, included the new module [into the docs](https://pydynamic.readthedocs.io/en/convolve_based_on_firuncfilter/PyDynamic.uncertainty.html#module-PyDynamic.uncertainty.propagate_convolution), suggested some minor docstring adjustments and some minor improvements for the method and the tests. I think we are ready to do a walk-through together. I am really keen on and looking forward on having explained what you do here.

File: PyDynamic/uncertainty/propagate_convolution.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+from .propagate_filter import _fir_filter
+
+__all__ = ["convolve_unc"]
+
+
+def convolve_unc(x1, U1, x2, U2, mode="full"):
+    """
+    An implementation of the discrete convolution of two signals with uncertainty propagation.
+    The code builds on PyDynamic.uncertainty._fir_filter, as convolution and filter application
```
[the remark this answers is not in the document]
Comment (mgrub) [reply PyDynamic/uncertainty/propagate_convolution.py:13]: If we would move `_fir_filter` to a public function, we would then have three functions (`fir_filter`, `fir_filter_diag`, `FIRuncFilter`. ) with very similar scope - which could to lead to confusion on the user-side. 
But the benefit of this would be, that we can then link this `fir_filter`-method in the corresponding documentations (``FIRuncFilter` and `convolve_unc`).

[the remark this answers is not in the document]
Comment (BjoernLudwigPTB) [reply PyDynamic/uncertainty/propagate_convolution.py:13]: The main question then would be, does the mentioning of `fir_filter` provide new insights or is that more of a technical detail? Or maybe even more: would a user in some cases rather call `fir_filter` directly?

File: PyDynamic/uncertainty/propagate_convolution.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    are mathematically nearly identical. However, boundary effects need to be taken into account.
+
+    Parameters
+    ----------
+    x1 : np.ndarray, (N,)
+        first input signal
+    U1 : np.ndarray, (N, N)
+        full 2D-covariance matrix associated with x1
+        if the signal is fully certain, use `U1 = None` (default) to make use of more efficient calculations.
```
[the remark this answers is not in the document]
Comment (mgrub) [reply PyDynamic/uncertainty/propagate_convolution.py:22]: Good spot. I removed the mentioning of the default-value alltogether. (If we wanted to set `U1` and `U2` to `None`, we would need to change the order of the arguments.)

File: PyDynamic/uncertainty/propagate_convolution.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    -------
+    y : np.ndarray
+        convoluted output signal
+    Uy : np.ndarray
+        full 2D-covariance matrix of y
+
+    References
+    ----------
+    .. seealso:: :mod:`numpy.convolve`
+    .. seealso:: :mod:`scipy.ndimage.convolve1d`
```
[the remark this answers is not in the document]
Comment (mgrub) [reply PyDynamic/uncertainty/propagate_convolution.py:48]: I changed that and also adjusted the structure of the `seealso::` usage.

File: PyDynamic/uncertainty/propagate_convolution.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+        if the signal is fully certain, use `U2 = None` (default) to make use of more efficient calculations.
+    mode : str, optional
+        numpy.convolve-modes:
+        - full:  len(y) == N+M-1 (default)
+        - valid: len(y) == max(M, N) - min(M, N) + 1
+        - same:  len(y) == max(M, N) (value+covariance are padded with zeros)
+        scipy.ndimage.convolve1d-modes:
+        - nearest: len(y) == N (value+covariance are padded with by stationary assumption)
+        - reflect:  len(y) == N
+        - mirror:   len(y) == N
```
[the remark this answers is not in the document]
Comment (mgrub) [reply PyDynamic/uncertainty/propagate_convolution.py:36]: 👍

File: PyDynamic/uncertainty/propagate_convolution.py [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+    # if a numpy-mode is chosen, x1 is expected to be the longer signal
+    # remeber that pure convolution is commutative
+    if len(x1) < len(x2) and mode in ["valid", "full", "same"]:
+        tmp = x1
+        x1 = x2
+        x2 = tmp
+
+        tmp = U1
+        U1 = U2
+        U2 = tmp
```
[the remark this answers is not in the document]
Comment (mgrub) [reply PyDynamic/uncertainty/propagate_convolution.py:61]: 🍬

File: PyDynamic/uncertainty/propagate_convolution.py [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+    .. seealso:: :mod:`scipy.ndimage.convolve1d`
+    .. seealso:: :mod:`PyDynamic.uncertainty.propagate_filter._fir_filter`
+    """
+
+    # if a numpy-mode is chosen, x1 is expected to be the longer signal
+    # remeber that pure convolution is commutative
+    if len(x1) < len(x2) and mode in ["valid", "full", "same"]:
+        tmp = x1
+        x1 = x2
+        x2 = tmp
```
[the remark this answers is not in the document]
Comment (mgrub) [reply PyDynamic/uncertainty/propagate_convolution.py:57]: I wasn't aware of that notation. That is really neat syntactic sugar 🤓 🍬

File: PyDynamic/uncertainty/propagate_convolution.py [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+
+    References
+    ----------
+    .. seealso:: :mod:`numpy.convolve`
+    .. seealso:: :mod:`scipy.ndimage.convolve1d`
+    .. seealso:: :mod:`PyDynamic.uncertainty.propagate_filter._fir_filter`
+    """
+
+    # if a numpy-mode is chosen, x1 is expected to be the longer signal
+    # remeber that pure convolution is commutative
```
[the remark this answers is not in the document]
Comment (mgrub) [reply PyDynamic/uncertainty/propagate_convolution.py:53]: 👍

File: test/test_propagate_convolution.py (hunk 3 above)
Comment (mgrub) [reply test/test_propagate_convolution.py:84]: Good point, I removed the `else`-branch.

File: test/test_propagate_convolution.py (hunk 4 above)
Comment (mgrub) [reply test/test_propagate_convolution.py:114]: 👍

File: test/test_propagate_convolution.py (hunk 1 above)
Comment (mgrub) [reply test/test_propagate_convolution.py:1]: Yes, that is true. I should some day look much more into what hypothesis already has to offer. 

But the random valid covariance-matrix is probably a bit too specific 😄 (The convolution-algorithm itself should be quite robust against non-plausible covariance inputs, but in the Monte-Carlo comparison `np.random.multivariate_normal` requires something plausible.)

File: PyDynamic/uncertainty/propagate_convolution.py [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+        )
+
+        # remove leading and trailing entries from output
+        conv = y[n2 : n2 + n1]
+        Uconv = Uy[n2 : n2 + n1, n2 : n2 + n1]
+
+    else:
+        raise ValueError(
+            'convolve_unc: Mode "{MODE}" is not supported.'.format(MODE=mode)
+        )
```
[the remark this answers is not in the document]
Comment (mgrub) [reply PyDynamic/uncertainty/propagate_convolution.py:152]: Thanks for the proposed test 👍

File: PyDynamic/uncertainty/propagate_convolution.py [hunk 13]
```diff
@@ -0,0 +1,147 @@
+# -*- coding: utf-8 -*-
+
```
[the remark this answers is not in the document]
Comment (BjoernLudwigPTB) [reply PyDynamic/uncertainty/propagate_convolution.py:10]: Has been added meanwhile.

File: PyDynamic/uncertainty/propagate_convolution.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (BjoernLudwigPTB) [reply PyDynamic/uncertainty/propagate_convolution.py:21]: So we adapted the docstring to keep it one abstraction layer higher and keep `_fir_filter` private.

File: PyDynamic/uncertainty/propagate_convolution.py [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+    y : np.ndarray
+        convoluted output signal
+    Uy : np.ndarray
+        full 2D-covariance matrix of y
+
+    References
+    ----------
+    .. seealso:: :mod:`numpy.convolve`
+    .. seealso:: :mod:`scipy.ndimage.convolve1d`
+    .. seealso:: :mod:`PyDynamic.uncertainty.propagate_filter._fir_filter`
```
[the remark this answers is not in the document]
Comment (BjoernLudwigPTB) [reply PyDynamic/uncertainty/propagate_convolution.py]: We removed the link after raising the docstring one abstraction layer higher.
